### PR TITLE
change order of early_stopping_rounds and run_options

### PR DESCRIPTION
**Pull request by ericlee0803**
Okay, so I have created a separate branch for hog/early_stopping, which I plan to merge into hog/v2_changes. hog/v2_changes will be squashed and merged into master when we deem it ready. 

Anyway, I talked to prod and they are okay with early stopping on a different metric than the optimization metric. It's much simpler to do so because we don't have to write any custom callbacks that compute accuracy, f1, precision, etc. 

I thought about it, and doing is going to turn into a headache when we consider support for xgboost internal metrics. We would need to understand which metrics are sigopt custom, and which metrics are xgboost internal, and handle these two cases separately (the former with a custom callback and the latter by adding the eval metric in question to the xgb parameter list). 

As it stands, if we accept early stopping without using the optimization metric, XGB defaults the early stopping metric to RMSE for regression, and logloss for classification, mean average precision for ranking. Which is fine with me.

### CODE REVIEW

**Comment by ericlee0803** [comment]
Yuhong could you take a look this week first? Harvey is a bit busy.

**Comment by yuhong-sigopt** [comment]
In run.py `log_default_params`, we may need also log `early_stopping_rounds`.

File: integration_test/test_xgboost_run.py [hunk 1]
```diff
@@ -146,6 +149,12 @@ def _verify_metric_logging(self, run):
 
     assert run.values['Training time'].value > 0
 
+    if 'early_stopping_rounds' in self.run_params:
+      assert run.values['num_boost_round_before_stopping']
+      if run.values['num_boost_round_before_stopping'].value < run.assignments['num_boost_round']:
```
Comment (yuhong-sigopt) [inline integration_test/test_xgboost_run.py:154]: Maybe the `<` should be `<=`

Comment (ericlee0803) [reply integration_test/test_xgboost_run.py:154]: I want to check if early stopping actually happened. Thus why I used <; if num_boost_round_before_stopping = num_boost_round then early stopping didn't do anything.

**Comment by ericlee0803** [comment]
Why do we need to log early_stopping_rounds in log_default_params? It's not activated by default, and therefore doesn't have a default value. I guess we could record None but that seems unnecessary

**Comment by yuhong-sigopt** [comment]
> Why do we need to log early_stopping_rounds in log_default_params? It's not activated by default, and therefore doesn't have a default value. I guess we could record None but that seems unnecessary

You are right.

File: integration_test/test_xgboost_run.py (hunk 1 above)
Comment (yuhong-sigopt) [reply integration_test/test_xgboost_run.py:154]: I think for some datasets, the train will stop before `early stop` takes effect.

File: integration_test/test_xgboost_experiment.py [hunk 2]
```diff
@@ -104,6 +104,20 @@ def test_experiment(self, task):
...
+    self.experiment_params['early_stopping_rounds'] = 1
+    experiment = sigopt.xgboost.experiment(**self.experiment_params)
+    assert experiment.is_finished()
+    sigopt_suggested_runs = list(experiment.get_runs())
+    for ssr in sigopt_suggested_runs:
+      assert ssr.assignments['early_stopping_rounds']
+      for param in experiment.parameters:
+        if param.name == 'num_boost_round':
+          assert ssr.values['num_boost_round_before_stopping']
+          assert ssr.assignments['num_boost_round'] >= ssr.assignments['num_boost_round_before_stopping']
```
[the remark this answers is not in the document]
Comment (harvey-intel) [reply integration_test/test_xgboost_experiment.py:119]: ^ If you test never caught this, that means this test is not very useful. My guess is that most of these runs would not be early stopped (because `num_boost_round` is very small already).

It's sufficient to check whether `early_stopping_round` is passed to the run level.

File: integration_test/test_xgboost_run.py [hunk 3]
```diff
@@ -146,6 +149,12 @@ def _verify_metric_logging(self, run):
...
+        assert run.values['best_iteration']
```
[the remark this answers is not in the document]
Comment (harvey-intel) [reply integration_test/test_xgboost_run.py:155]: You named this `Best iteration` in the actual code.


Same comment regarding your test not being able to catch this error.

File: sigopt/xgboost/run.py [hunk 4]
```diff
@@ -278,10 +280,15 @@ def train_xgb(self):
...
     # Always log xgb-default eval_metric
+    n_eval_rounds = 0
     if self.evals_result is not None:
       for dataset, metric_dict in self.evals_result.items():
         for metric_label, metric_record in metric_dict.items():
           self.run.log_metric(f"{dataset}-{metric_label}", metric_record[-1])
+          n_eval_rounds = len(metric_record)
+
+      if self.early_stopping_rounds:
+        self.run.log_metric('num_boost_round_before_stopping', n_eval_rounds)
```
Comment (harvey-intel) [inline sigopt/xgboost/run.py:291]: `num_boost_round_before_stopping` has underscores; `Best iteration` has no underscore. Make them consistent.

File: integration_test/test_xgboost_run.py (hunk 1 above)
Comment (harvey-intel) [reply integration_test/test_xgboost_run.py:154]: then you should also check the else statement, in which case there is no `best_iteration` metric output.

File: integration_test/test_xgboost_run.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (harvey-intel) [reply integration_test/test_xgboost_run.py:155]: What makes sense is to come up with a "trivial" dataset  where you know xgboost is guaranteed to early stop... and then test all your corresponding logging stuff.

**Review by harvey-intel — CHANGES_REQUESTED** [review]
Some minor style comments in the actual implementation code.

Need to rethink on how to test early stopping more directly. The test code right now is probably being activated at a very low rate; therefore not very helpful.

File: integration_test/test_xgboost_run.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ericlee0803) [reply integration_test/test_xgboost_run.py:155]: Good catch! Thanks. And yes, I will do that.

File: integration_test/test_xgboost_experiment.py [hunk 5]
```diff
@@ -104,6 +104,20 @@ def test_experiment(self, task):
...
 
+  def test_early_stopping(self):
+    task = 'binary'
+    self._form_random_experiment_config(task)
+    self.experiment_params['early_stopping_rounds'] = 1
+    experiment = sigopt.xgboost.experiment(**self.experiment_params)
+    assert experiment.is_finished()
+    sigopt_suggested_runs = list(experiment.get_runs())
+    for ssr in sigopt_suggested_runs:
+      assert ssr.assignments['early_stopping_rounds']
```
[the remark this answers is not in the document]
Comment (ericlee0803) [reply integration_test/test_xgboost_experiment.py:115]: I guess it's a little tricky. It's not explicitly a parameter that's optimized, but it's being logged as a parameter in the runs API, but with source "User Specified"

[the remark this answers is not in the document]
Comment (ericlee0803) [reply integration_test/test_xgboost_experiment.py:115]: I guess we could test for the source too, not sure how to do that at the moment. @yuhong-sigopt do you know how to do this?

**Comment by ericlee0803** [comment]
Edit: despite what the XGB docs say, bst.best_iteration is saved every time a xgb model is trained, not just when early stopping has occurred. So I've made the change to always log bst.best_iteration (unless auto logging metrics is off).

File: integration_test/test_xgboost_experiment.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (harvey-intel) [reply integration_test/test_xgboost_experiment.py:115]: I see, it's an assignment associated with the run. then can you just
`assert ssr.assignments['early_stopping_rounds'] == self.experiment_param['early_stopping_rounds']`

**Comment by ericlee0803** [comment]
I added an additional test, which fixes the random seed and generates random X data. Then, Y data is created to be mostly zero and early_stopping_rounds is set to 2, which causes XGB to stop almost immediately. 

Using this example that is known to early stop, I then test that num_boost_round_before_stopping < num_boost_round.

File: integration_test/test_xgboost_run.py [hunk 6]
```diff
@@ -365,3 +367,27 @@ def test_xgbrun_callbacks_no_appending(self):
...
+    d = 10
+    n = 500
+    numpy.random.seed(1234)
+    X = numpy.random.rand(n, d)
+    y = numpy.zeros(n)
+    y[0:10] = 1
+    X_train, X_test, Y_train, Y_test = train_test_split(X, y, test_size=0.2, random_state=1234)
+    D_train = xgb.DMatrix(X_train, label=Y_train)
+    D_test = xgb.DMatrix(X_test, label=Y_test)
+    self.run_params = _form_random_run_params(task="binary")
```
Comment (harvey-intel) [inline integration_test/test_xgboost_run.py:385]: technically don't need this one, can just be `self.run_params = {}`.

File: integration_test/test_xgboost_experiment.py [hunk 7]
```diff
@@ -104,6 +104,19 @@ def test_experiment(self, task):
...
+  def test_early_stopping(self):
+    self._form_random_experiment_config('binary')
+    self.experiment_params['early_stopping_rounds'] = 1
+    experiment = sigopt.xgboost.experiment(**self.experiment_params)
+    assert experiment.is_finished()
+    sigopt_suggested_runs = list(experiment.get_runs())
+    for ssr in sigopt_suggested_runs:
+      assert ssr.assignments['early_stopping_rounds'] == self.experiment_params['early_stopping_rounds']
+      for param in experiment.parameters:
+        if param.name == 'num_boost_round':
```
Comment (harvey-intel) [inline integration_test/test_xgboost_experiment.py:116]: Does this for loop and if statement need to exist? If early stopping happens, that's regardless of whether we are setting `num_boost_round` as a tuning parameter or not, no?

**Review by harvey-intel — COMMENTED** [review]
Some nit picks. mostly looks good.

File: sigopt/xgboost/experiment.py [hunk 8]
```diff
@@ -181,14 +183,31 @@ def run_experiment(self):
...
 
 
-def experiment(experiment_config, dtrain, evals, params, num_boost_round=None, run_options=None):
+def experiment(
+  experiment_config,
+  dtrain,
+  evals,
+  params,
+  num_boost_round=None,
+  early_stopping_rounds=10,
```
Comment (harvey-intel) [inline sigopt/xgboost/experiment.py:198]: Since `early_stopping_rounds` is now set by us, we should label this parameter source as `sigopt default`.

Comment (ericlee0803) [reply sigopt/xgboost/experiment.py]: I added a new source called SigOpt default and am recording early stopping rounds using this source

File: sigopt/xgboost/experiment.py [hunk 9]
```diff
@@ -181,14 +186,42 @@ def run_experiment(self):
           self.dtrain,
           num_boost_round=num_boost_round_run,
           evals=self.evals,
+          early_stopping_rounds=self.early_stopping_rounds,
           verbose_eval=False,
           run_options=self.run_options,
         )
 
+        # mark early stopping rounds as SigOpt Default
+        if self.early_stopping_rounds == DEFAULT_EARLY_STOPPING_ROUNDS:
```
Comment (harvey-intel) [inline sigopt/xgboost/experiment.py:195]: Should also check that `early_stopping_rounds` is not defined in the `experiment_config["parameters"]`, or better yet, forbid users from tuning this parameter altogether.

File: sigopt/xgboost/experiment.py [hunk 10]
```diff
@@ -181,14 +186,42 @@ def run_experiment(self):
...
+          run.set_parameters_sources_meta(
+            SIGOPT_DEFAULTS_SOURCE_NAME,
+            sort=SIGOPT_DEFAULTS_SOURCE_PRIORITY,
+            default_show=True
```
Comment (harvey-intel) [inline sigopt/xgboost/experiment.py:199]: trailing comma.

File: sigopt/xgboost/constants.py [hunk 11]
```diff
@@ -1,15 +1,18 @@
 DEFAULT_EVALS_NAME = 'TestSet'
 USER_SOURCE_NAME = 'User Specified'
+SIGOPT_DEFAULTS_SOURCE_NAME = 'SigOpt Defaults'
 XGBOOST_DEFAULTS_SOURCE_NAME = 'XGBoost Defaults'
 
 USER_SOURCE_PRIORITY = 2
-XGBOOST_DEFAULTS_SOURCE_PRIORITY = 3
+SIGOPT_DEFAULTS_SOURCE_PRIORITY = 3
+XGBOOST_DEFAULTS_SOURCE_PRIORITY = 4
```
Comment (harvey-intel) [inline sigopt/xgboost/constants.py:8]: Reordering this might screw up the web...cc @simonhowey ?

**Review by harvey-intel — APPROVED** [review]
A few small checks, but mostly good to go.

File: sigopt/xgboost/constants.py (hunk 11 above)
Comment (ericlee0803) [reply sigopt/xgboost/constants.py:8]: I'll defer to Simon on this. At the very least, it seems to be rendering okay: 
https://app.sigopt.com/run/193376

Comment (harvey-intel) [reply sigopt/xgboost/constants.py:8]: Would this mess up at the Project level?

Comment (simonhowey) [reply sigopt/xgboost/constants.py:8]: Should be fine the relative number is what matters not the actual number

File: sigopt/xgboost/experiment.py (hunk 9 above)
Comment (simonhowey) [reply sigopt/xgboost/experiment.py:195]: Also want to ensure they aren't passing in same default as us. Seems like 10 might be reasonably common?

https://stackoverflow.com/questions/67202314/how-to-distinguish-default-argument-and-argument-provided-with-default-value

File: sigopt/xgboost/constants.py (hunk 11 above)
Comment (harvey-intel) [reply sigopt/xgboost/constants.py:8]: I guess we only need to update the coloring scheme on the web dashboard then.

File: sigopt/xgboost/experiment.py (hunk 9 above)
Comment (harvey-intel) [reply sigopt/xgboost/experiment.py:195]: good catch.

Comment (ericlee0803) [reply sigopt/xgboost/experiment.py]: Thanks for catching this, just updated and re-requested review.
